lib/modules/platform/bitbucket-server/utils.ts

```diff
@@ -155,39 +157,81 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
   return invalidReviewers;
 }
 
+function generateUrlFromEndpoint(
+  defaultEndpoint: string,
+  opts: HostRule,
+  repository: string
+): string {
+  const url = new URL(defaultEndpoint);
+  const generatedUrl = git.getUrl({
+    protocol: url.protocol as GitProtocol,
+    auth: `${opts.username}:${opts.password}`,
+    host: `${url.host}${url.pathname}${
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+      url.pathname!.endsWith('/') ? '' : /* istanbul ignore next */ '/'
+    }scm`,
+    repository,
+  });
+  logger.debug({ url: generatedUrl }, `using generated endpoint URL`);
+  return generatedUrl;
+}
+
 export function getRepoGitUrl(
   repository: string,
   defaultEndpoint: string,
+  gitUrl: GitUrlOption | undefined,
```

Comment:
this function can be simplified, use a switch case on `gitUrl` and throw a config error if `gitUrl` is set, but type couldn't found ( http, ssh )

Reply:
Do you mean something like:
`  switch(gitUrl) {
    case "default": {
      // do something;
      break;
    }
    case "ssh": {
      // do something;
      break;
    }
    case "endpoint": {
      // do something;
      break;
    }
    case undefined: {
      // do something;
      break;
    }
    default: {
      //throw error;
      break;
    }
  }`

Reply:
Changes applied. 

I used undefined to follow the existing logic where we try http then ssh and finally generate the URL.